### PR TITLE
Added support for server shutting down state

### PR DESCRIPTION
**Pull request by snarkychef**
- Added rruntime package to listen on terminate signals
- Gateway responds 503 if server shutdown is in progress
- Refactored stats and profiling packages into monitoring package

---
This change is [Reviewable](https://reviewable.io/reviews/rudderlabs/rudder-server/134)

### CODE REVIEW

**Comment by snarkychef** [comment]
WIP don't merge yet

File: gateway/gateway.go [hunk 1]
```diff
@@ -265,7 +267,7 @@ func (gateway *HandleT) webRequestBatchDBWriter(process int) {
 				misc.IncrementMapByKey(writeKeySuccessStats, jobWriteKeyMap[uuid], 1)
 				misc.IncrementMapByKey(writeKeySuccessEventStats, jobWriteKeyMap[uuid], jobEventCountMap[uuid])
 			}
-			jobIDReqMap[uuid].done <- err
+			jobIDReqMap[uuid].done <- DBPersistFailed
```
Comment (lokey) [inline gateway/gateway.go]: What if err is not DBPersistFailed?

File: router/batchrouter/batchrouter.go [hunk 2]
```diff
@@ -269,12 +270,12 @@ func (brt *HandleT) setJobStatus(batchJobs BatchJobsT, isWarehouse bool, err err
 func (brt *HandleT) initWorkers() {
 	for i := 0; i < noOfWorkers; i++ {
 		go func() {
-			for {
+			for !rruntime.IsShutDownInProgess {
```
Comment (lokey) [inline router/batchrouter/batchrouter.go]: 👍

**Comment by snarkychef** [comment]
Please use feature spec in the below mentioned link for performing the review. 
https://app.gitbook.com/@rudderlabs/s/internal-docs/~/drafts/-M-PKE8QlEbHpkcCQRsa/backend-features-spec/clean-shutdown

File: gateway/gateway.go (hunk 1 above)
Comment (snarkychef) [reply gateway/gateway.go]: Let's just log the err and not send specific error as part of client response.

**Pull request closed without merging**